ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -172,9 +173,18 @@ public void appendFrom(ReadableByteChannel channel, long size) throws IOExceptio
           "Channel cannot be written to segment [" + file.getAbsolutePath() + "] because" + " it exceeds the capacity ["
               + capacityInBytes + "]");
     } else {
+      if (!fileChannel.isOpen()) {
+        throw new ClosedChannelException();
+      }
       long bytesWritten = 0;
-      while (bytesWritten < size) {
-        bytesWritten += fileChannel.transferFrom(channel, endOffset.get() + bytesWritten, size - bytesWritten);
+      long bytesRead = 0;
+      ByteBuffer bytebuffer = ByteBuffer.allocateDirect((int) size);
```

Comment:
As we talked about offline, you can use https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html to create a separate buffer for each thread.